packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart

```diff
@@ -153,9 +154,45 @@ class DeepLinksController extends DisposableController
         selectedIosConfigurationIndex,
         _handleIosConfigurationChanged,
       );
+      addAutoDisposeListener(
+        selectedIosTargetIndex,
+        _handleIosConfigurationChanged,
+      );
+    }
+  }
+
+  void firstLoadWithDefaultConfigurations() async {
+    firstLoadWithDefaultConfigurationsCompleted = false;
```

Comment:
instead of setting this to false, should we exit early? it is possible that this method will be called again when the `_DeepLinkListViewState` is disposed, and later recreated.

Consider this approach, which could prevent you from having to track a firstLoad boolean:

```
DeepLinksController() {
  unawaited(_init());
}

// You can add these if you need any other code to wait on
// the initialization of the DeepLinksController.
Future<void> get initialized => _initialized.future;
final _initialized = Completer<void>();

Future<void> _init() {
    // Set all the default indices

    selectedAndroidVariantIndex.value = _getDefaultConfigurationIndex(
      selectedProject.value!.androidVariants,
      containsString: 'release',
    );
    if (FeatureFlags.deepLinkIosCheck) {
      selectedIosConfigurationIndex.value = _getDefaultConfigurationIndex(
        selectedProject.value!.iosBuildOptions.configurations,
        containsString: 'release',
      );
      selectedIosTargetIndex.value = _getDefaultConfigurationIndex(
        selectedProject.value!.iosBuildOptions.configurations,
        containsString: 'runner',
      );
    }
    await loadLinksAndValidate();

    // Then add the listeners so they aren't triggered by the above code.

    addAutoDisposeListener(
      selectedAndroidVariantIndex,
      _handleAndroidConfigurationChanged,
    );
    if (FeatureFlags.deepLinkIosCheck) {
      addAutoDisposeListener(
        selectedIosConfigurationIndex,
        _handleIosConfigurationChanged,
      );
      addAutoDisposeListener(
        selectedIosTargetIndex,
        _handleIosConfigurationChanged,
      );
    }

    _initialized.complete();
}
```

This will guarantee that all this initialization code only gets called once, independent of the widget lifecycle of the `DeepLinkListView` widget. You may need to handle the case where `selectedProject.value!` may be null. It is possible I am missing some contextual understanding of this feature though, so feel free to push back if you intended this code to be called more than once.

Reply:
DeepLinksController is initialized first, and then user select a project and selectedProject.value is set, So I can’t use selectedProject.value in DeepLinksController’s initialization. 

I can update the DeepLinksController into 2 controller to do this.
DeepLinksScreenController<DeepLinksScreen<DeepLinkListViewController<DeepLinkListView>>>
And DeepLinksScreenController only controls the variable of selectedProject, DeepLinkListViewController controls everything else. 
Do you think it makes sense?

Reply:
Okay after looking at it a little deeper, I think the pattern that we should change here is that the selection notifiers expose the ValueNotifier to the UI instead of just exposing a ValueListenable. If the notifiers are private, then all changes to the notifiers must go through the controller, which gives us a bit more control over how we handle changes to the selections. The fact that the controller is listening to notifiers that it owns is a little strange. 

So I would recommend a pattern like this:

```
class DeepLinksController {
  ValueListenable<int> get selectedAndroidVariantIndex => _selectedAndroidVariantIndex;
  final _selectedAndroidVariantIndex = ValueNotifier<int>(0);
  void updateSelectedAndroidVariantIndex(int index) {
    _selectedAndroidVariantIndex.value = index;
    _handleAndroidConfigurationChanged(); 
  }
}
```

And then your UI code should listen to changes from the `selectedAndroidVariantIndex` ValueListenable, and update the value by calling `controller.updateSelectedAndroidVariantIndex`.

Make sense?

Reply:
You're right, I was spiraling about listening to controller's own notifier and it's a strange pattern.

The new pattern is much clear!